connect/api/src/main/java/org/apache/kafka/connect/storage/SimpleHeaderConverter.java

```diff
@@ -0,0 +1,728 @@
+        } else if (value instanceof Boolean) {
+            sb.append(value);
+        } else if (value instanceof String) {
+            if (embedded) {
+                String escaped = escape((String) value);
+                sb.append('"').append(escaped).append('"');
+            } else {
+                sb.append(value);
+            }
+        } else if (value instanceof byte[]) {
+            // Same as StringConverter
+            if (embedded) {
+                sb.append('"').append(value).append('"');
+            } else {
+                sb.append(value);
+            }
+        } else if (value instanceof List) {
+            List<?> list = (List<?>) value;
+            sb.append('[');
+            appendIterable(sb, list.iterator());
+            sb.append(']');
+        } else if (value instanceof Map) {
+            Map<?, ?> map = (Map<?, ?>) value;
+            sb.append('{');
+            appendIterable(sb, map.entrySet().iterator());
+            sb.append('}');
+        } else if (value instanceof Struct) {
+            Struct struct = (Struct) value;
+            Schema schema = struct.schema();
+            boolean first = true;
+            sb.append('{');
+            for (Field field : schema.fields()) {
+                if (first) {
+                    first = false;
+                } else {
+                    sb.append(',');
+                }
+                append(sb, struct.get(field), true);
+            }
+            sb.append('}');
+        } else if (value instanceof Map.Entry) {
+            Map.Entry<?, ?> entry = (Map.Entry<?, ?>) value;
+            append(sb, entry.getKey(), true);
+            sb.append(':');
+            append(sb, entry.getValue(), true);
+        } else if (value instanceof java.util.Date) {
+            java.util.Date dateValue = (java.util.Date) value;
+            String formatted = dateFormatFor(dateValue).format(dateValue);
+            sb.append(formatted);
+        } else {
+            throw new DataException("Failed to serialize unexpected value type " + value.getClass().getName() + ": " + value);
+        }
+    }
+
+    protected void appendIterable(StringBuilder sb, Iterator<?> iter) {
+        if (iter.hasNext()) {
+            append(sb, iter.next(), true);
+            while (iter.hasNext()) {
+                sb.append(',');
+                append(sb, iter.next(), true);
+            }
+        }
+    }
+
+    protected DateFormat dateFormatFor(java.util.Date value) {
+        if (value.getTime() < MILLIS_PER_DAY) {
+            return timeFormatter;
+        }
+        if (value.getTime() % MILLIS_PER_DAY == 0) {
+            return dateFormatter;
+        }
+        return timestampFormatter;
+    }
+
+    protected String escape(String value) {
+        return value.replaceAll("\\\\", "\\\\\\\\").replaceAll("\"", "\\\\\"");
+    }
+
+    protected SchemaAndValue parse(Tokenizer tokenizer, boolean embedded) throws NoSuchElementException {
```

Comment:
I am a bit worried that this parser could change the underlying data in small ways, and cause issues which become hard to debug. For example, the string "  0.11" would drop the initial spaces, and numbers in scientific notation might be translated into decimal (for example, 1E-2 becomes 0.01). I think this class is very useful, but do you think we want to have `org.apache.kafka.connect.converters.ByteArrayConverter` as the default converter?

Reply:
Okay, what if we move all of this conversion logic into the `Header.valueAsX` methods? If we do that, I think we can use the `StringConverter` as the default and can avoid the `SimpleHeaderConverter` class altogether.

Reply:
Yes, that would make sense. The transformations would be explicit. Should we use `ByteArrayConverter` as the default?

Reply:
After thinking about this more, the existing `StringConverter` uses `toString()` to serialize all objects, which means for arrays and maps the string representation contains no quotes around strings and so would be relatively difficult to parse. So I'm not convinced we can do without `SimpleHeaderConverter`.

Also, I'm not sure why it'd be an issue if the default header converter parser the values into nominal values, even if that results in slight variations. First, existing connectors don't handle headers now. Second, any connector that did require the exact representation can specify a non-default header converter. Third, it makes more sense to me to make the headers meaningful and useful by default.

Reply:
I think there is a lot of value in providing meaning to the headers. If we could make the transition explicit, that would be the best. Since connect sits between many different systems with non compatible data models, a translation may cause unexpected issues in production. For example, a long in Java is up to 2^64-1, but javascript would use double precision to handle big integers.

Regarding the problem of serializing maps, I agree with you (`StringConverter` and `ByteArrayConverter` won't work). 

Would it fine to think that if people are converting headers in their tasks or SMTs, then we can expect them to set `SimpleHeaderConverter`  as their converter class? And if they never care about what is done to the headers, then it should be correct to have set `ByteArrayConverter` as the default value for `header.converter` property in `WorkerConfig.java`?

Reply:
Yes, it is certainly possible to use the `ByteArrayConverter` in either the worker config or the connector config, and when this is used the byte array in each header value will simply be passed directly into the `Headers` object for each record.

I'm still not convinced that this should be the *default* header converter, though.